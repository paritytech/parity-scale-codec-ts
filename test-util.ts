/// <reference lib="deno.unstable"/>

import { dirname, join } from "std/path/win32.ts";
import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { Codec } from "./common.ts";

const file = (path: string) => () => fetch(join(dirname(import.meta.url), path)).then((x) => x.text());
export const files = {
  lipsum: file("lipsum.txt"),
  words: file("words.txt"),
  cargoLock: file("Cargo.lock"),
};

export function testCodec<T>(name: string, codec: Codec<T>, values: T[] | Record<string, T>) {
  for (const key in values) {
    const value = values[key as never] as T;
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key;
    Deno.test(`${name} ${label}`, async (t) => {
      const encoded = codec.encode(value);
      const decoded = codec.decode(encoded);
      assertEquals(decoded, value);
      await assertSnapshot(t, encoded, { serializer: serializeU8A });
    });
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n");
}

export function benchCodec<T>(name: string, codec: Codec<T>, value: T) {
  const encoded = codec.encode(value);
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    codec.encode(value);
  });
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    codec.decode(encoded);
  });
}
