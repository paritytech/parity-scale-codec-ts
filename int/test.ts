import * as $ from "../mod.ts";
import { testCodec } from "../test-util.ts";

testCodec("u8", $.u8, [0, 1, 2, 9]);
testCodec("u16", $.u16, [0, 1, 2, 9]);
testCodec("u32", $.u32, [0, 1, 2, 9]);
testCodec("u64", $.u64, [0n, 1n, 2n, 9n]);
testCodec("u128", $.u128, [0n, 1n, 2n, 9n]);

testCodec("i8", $.i8, [-9, -2, -1, 0, 1, 2, 9]);
testCodec("i16", $.i16, [-9, -2, -1, 0, 1, 2, 9]);
testCodec("i32", $.i32, [-9, -2, -1, 0, 1, 2, 9]);
testCodec("i64", $.i64, [-9n, -2n, -1n, 0n, 1n, 2n, 9n]);
testCodec("i128", $.i128, [-9n, -2n, -1n, 0n, 1n, 2n, 9n]);
