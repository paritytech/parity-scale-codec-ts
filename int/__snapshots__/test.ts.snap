export const snapshot = {};

snapshot[`u8 0 1`] = `00`;

snapshot[`u8 1 1`] = `01`;

snapshot[`u8 2 1`] = `02`;

snapshot[`u8 9 1`] = `09`;

snapshot[`u16 0 1`] = `
00
00
`;

snapshot[`u16 1 1`] = `
01
00
`;

snapshot[`u16 2 1`] = `
02
00
`;

snapshot[`u16 9 1`] = `
09
00
`;

snapshot[`u32 0 1`] = `
00
00
00
00
`;

snapshot[`u32 1 1`] = `
01
00
00
00
`;

snapshot[`u32 2 1`] = `
02
00
00
00
`;

snapshot[`u32 9 1`] = `
09
00
00
00
`;

snapshot[`u64 0n 1`] = `
00
00
00
00
00
00
00
00
`;

snapshot[`u64 1n 1`] = `
01
00
00
00
00
00
00
00
`;

snapshot[`u64 2n 1`] = `
02
00
00
00
00
00
00
00
`;

snapshot[`u64 9n 1`] = `
09
00
00
00
00
00
00
00
`;

snapshot[`u128 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`u128 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`u128 2n 1`] = `
02
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`u128 9n 1`] = `
09
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`i8 -9 1`] = `f7`;

snapshot[`i8 -2 1`] = `fe`;

snapshot[`i8 -1 1`] = `ff`;

snapshot[`i8 0 1`] = `00`;

snapshot[`i8 1 1`] = `01`;

snapshot[`i8 2 1`] = `02`;

snapshot[`i8 9 1`] = `09`;

snapshot[`i16 -9 1`] = `
f7
ff
`;

snapshot[`i16 -2 1`] = `
fe
ff
`;

snapshot[`i16 -1 1`] = `
ff
ff
`;

snapshot[`i16 0 1`] = `
00
00
`;

snapshot[`i16 1 1`] = `
01
00
`;

snapshot[`i16 2 1`] = `
02
00
`;

snapshot[`i16 9 1`] = `
09
00
`;

snapshot[`i32 -9 1`] = `
f7
ff
ff
ff
`;

snapshot[`i32 -2 1`] = `
fe
ff
ff
ff
`;

snapshot[`i32 -1 1`] = `
ff
ff
ff
ff
`;

snapshot[`i32 0 1`] = `
00
00
00
00
`;

snapshot[`i32 1 1`] = `
01
00
00
00
`;

snapshot[`i32 2 1`] = `
02
00
00
00
`;

snapshot[`i32 9 1`] = `
09
00
00
00
`;

snapshot[`i64 -9n 1`] = `
f7
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i64 -2n 1`] = `
fe
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i64 -1n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i64 0n 1`] = `
00
00
00
00
00
00
00
00
`;

snapshot[`i64 1n 1`] = `
01
00
00
00
00
00
00
00
`;

snapshot[`i64 2n 1`] = `
02
00
00
00
00
00
00
00
`;

snapshot[`i64 9n 1`] = `
09
00
00
00
00
00
00
00
`;

snapshot[`i128 -9n 1`] = `
f7
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i128 -2n 1`] = `
fe
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i128 -1n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`i128 0n 1`] = `
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`i128 1n 1`] = `
01
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`i128 2n 1`] = `
02
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;

snapshot[`i128 9n 1`] = `
09
00
00
00
00
00
00
00
00
00
00
00
00
00
00
00
`;
